### PR TITLE
Loading Data for Form Fields

### PR DESCRIPTION
**Pull request by abdullai-t**
####  Summary / Highlights
Our dropdown form fields, which rely on data in Redux, currently display only a limited number of options for users to select from. However, there might be more data available in our database that is not being shown. This limitation is a result of the pagination implementation we have in place, both on the backend and within our application. Unfortunately, we overlooked updating our dropdown fields to allow fetching more data when the list of items is exhausted. This PR fixes this issue by making the dropdowns pull data from the db whenever the user reaches the bottom of the list.

### What Was Done
1. The formGenerator does not refresh its fields whenever there are changes to its items from outside. To address this issue, the `getDerivedStateFromProps` lifecycle method was utilised to listen for changes in the props and trigger a re-render when necessary.
2.  The `Dropdown`, `AutoComplete` and the `Checkbox` fields of the formGenerator now accept `isAsync`, `endpoint` and `labelExtractor` as additional properties. This properties should only be added if you want it to fetch data from the backend whenever the scroll reaches the bottom. The `labelExtractor` function is the same as the one used in reconstructing data for `AutoComplete` field.
3. This implementation was extended to all form dropdowns that lists actions, communities and vendors.

#### Demo

https://github.com/massenergize/frontend-admin/assets/42780120/756422ef-6605-4d67-a4b6-59019c58a768

### CODE REVIEW

**Review by frimpongopoku — CHANGES_REQUESTED** [review]
Well done @abdullai-t  🥳 . 
Maybe in future, we can make the intersection observer logic you have implemented so many times a component on its own? So that we can use it to wrap any set of nodes and make it do something when a particular node is in view...?



And please make sure the dropdown item paddings are consistent... seee here: 
Screenshot 2023-05-30 at 09 25 52
Screenshot 2023-05-30 at 09 26 12


That said, everything else LGTM 🤝 🎊

File: app/containers/MassEnergizeSuperAdmin/Teams/CreateNewTeamForm.js [hunk 1]
```diff
@@ -46,19 +46,21 @@ class CreateNewTeamForm extends Component {
...
   static getDerivedStateFromProps(props, state) {
-    var { communities, formState, location } = props;
+    var { communities, formState, location, auth } = props;
     communities = (communities || []).map((c) => ({
       ...c,
       displayName: c.name,
-      id: "" + c.id,
+      id:c.id,
     }));
+      const isSuperAdmin = auth?.is_super_admin && !auth?.is_community_admin;
```
Comment (BradHN1) [inline app/containers/MassEnergizeSuperAdmin/Teams/CreateNewTeamForm.js:55]: This doesn't look correct.   isSuperAdmin should override isCommunityAdmin.  So if you are both, the admin portal should treat you as superAdmin everywhere.

File: app/containers/MassEnergizeSuperAdmin/Testimonials/CreateNewTestimonialForm.js [hunk 2]
```diff
@@ -44,33 +44,33 @@ class CreateNewTestimonialForm extends Component {
...
-    var { vendors, actions, tags, communities } = props;
+    var { vendors, actions, tags, communities, auth } = props;
     const readyToRenderThePageFirstTime =
       vendors &&
       actions &&
       actions.length &&
       tags &&
       tags.length;
-
+    const isSuperAdmin = auth?.is_super_admin && !auth?.is_community_admin;
```
Comment (BradHN1) [inline app/containers/MassEnergizeSuperAdmin/Testimonials/CreateNewTestimonialForm.js:54]: same comment as above.  is_super_admin should supercede is_community_admin

File: app/containers/MassEnergizeSuperAdmin/Testimonials/EditTestimonial.js [hunk 3]
```diff
@@ -71,24 +71,25 @@ class EditTestimonial extends Component {
       actions.length&& 
       tags &&
       tags.length;
+      const isSuperAdmin = auth?.is_super_admin && !auth?.is_community_admin;
```
Comment (BradHN1) [inline app/containers/MassEnergizeSuperAdmin/Testimonials/EditTestimonial.js:74]: is_super_admin supersedes is_community_admin

File: app/containers/MassEnergizeSuperAdmin/Vendors/AddVendorForm.js [hunk 4]
```diff
@@ -46,7 +46,10 @@ class CreateNewVendorForm extends Component {
   }
 
   static getDerivedStateFromProps(props, state) {
-    const { communities, tags, formState, location } = props;
+    const { communities, tags, formState, location, auth } = props;
+
+      const isSuperAdmin =
+        auth?.is_super_admin && !auth?.is_community_admin;
```
Comment (BradHN1) [inline app/containers/MassEnergizeSuperAdmin/Vendors/AddVendorForm.js:52]: is_super_admin overrules is_community_admin

File: app/containers/MassEnergizeSuperAdmin/Vendors/EditVendor.js [hunk 5]
```diff
@@ -49,7 +49,9 @@ class CreateNewVendorForm extends Component {
 
   static getDerivedStateFromProps(props, state) {
     // you need: communities, vendor, vendors, tags
-    const { match, communities, vendors, tags, vendorsInfos, location } = props;
+    const { match, communities, vendors, tags, vendorsInfos, location, auth } = props;
+          const isSuperAdmin =
+            auth?.is_super_admin && !auth?.is_community_admin;
```
Comment (BradHN1) [inline app/containers/MassEnergizeSuperAdmin/Vendors/EditVendor.js:54]: same

**Pull request merged**